### PR TITLE
Fix #491: Zip and tar files upload size limit

### LINKED ISSUE #491: Zip file upload size limit
Apparently, the limit that we enforce on individual file sizes for upload (3mb) is also enforced for zip files that we expect to unzip inside the file tree i.e. I cannot upload a zip file that is > 3mb.

I don't think that this should be the case because the zip file can contain multiple < 3mb files that would be perfectly valid to upload individually. Seems like a bit of an unnecessary UX hassle to do it that way though.

### PR DESCRIPTION
**Pull request by MargarytaChepiga**
Hello, here is the link to the issue itself: https://github.com/mozilla/brackets/issues/491
I changed the rejectImport function (https://github.com/mozilla/brackets/blob/master/src/utils/DragAndDrop.js#L330-L349), so it's allowing to import tar and zip files up to 5MB. Note that maximum size for these extensions can be easily changed. So for example, we can change it to 10MB ( since the size is a variable, like the byteLimit for all other files).
here is a gif image to represent the fix:
bug

Here is a gif image representing that we can't upload any kind of file that is bigger that 5MB:
bug_with_error

Here is a gif image representing that we can't upload any kind of file bigger that 3MB (beside zip or tar):
bug_test

While I was doing this bug, I also noticed issue 1 and issue 2. Issue 1 I’ve filed in #1833 (https://github.com/mozilla/thimble.mozilla.org/issues/1833). However issue 2 is caused by my code. 
bug_with_error

Basically, the bug is in the error message. For files that are not zip or tar, the max size is 3MB, therefore, the error message for them is correct. However, for tar and zip files, the max size is now 5MB, but the message tells is't 3MB. 

I am going to create an issue on this bug, if my pull request will be accepted, otherwise, this issue does not exist.

### CODE REVIEW

File: src/utils/DragAndDrop.js [hunk 1]
```diff
@@ -332,13 +335,21 @@ define(function (require, exports, module) {
          * or not a mime type we care about, reject it.
          */
         function rejectImport(item) {
-            if (item.size > byteLimit) {
+            var ext = Path.extname(item.name).replace(/^\./, "").toLowerCase();
+            if (item.size > archiveByteLimit) {
                 return new Error(Strings.DND_MAX_FILE_SIZE_EXCEEDED);
+            }
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: Please move your `else if` to the same line as your closing `}`, should be `} else if {`

File: src/utils/DragAndDrop.js [hunk 2]
```diff
@@ -332,13 +335,21 @@ define(function (require, exports, module) {
...
+                    return new Error(Strings.DND_MAX_FILE_SIZE_EXCEEDED);
+                }
             }
 
+            
+
             // If we don't know about this language type, or the OS doesn't think
             // it's text, reject it.
-            var ext = Path.extname(item.name).replace(/^\./, "").toLowerCase();
+            //var ext = Path.extname(item.name).replace(/^\./, "").toLowerCase();
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: You can just delete this line altogether, since you're doing it above.

File: src/utils/DragAndDrop.js [hunk 3]
```diff
@@ -332,13 +335,17 @@ define(function (require, exports, module) {
          * or not a mime type we care about, reject it.
          */
         function rejectImport(item) {
-            if (item.size > byteLimit) {
+            var ext = Path.extname(item.name).replace(/^\./, "").toLowerCase();
```
Comment (Pomax) [inline src/utils/DragAndDrop.js]: does this actually require a regexp? Can we not use the standard `replace(".","")` given the output of `extname()`?

Comment (humphd) [inline src/utils/DragAndDrop.js]: Indeed, we could, and @MargarytaChepiga is doing a good job following the style already in that file--in another place we do it this way.

Comment (Pomax) [inline src/utils/DragAndDrop.js]: Fair point. looking at the rest of the code there are also a fair number of spots where file type checks based on extension run through `Content.is[...](Path.extname(filename))`, does it make sense to extend Content here so that it has an `isArchive()` function that accepts .tar, .zip, etc?

Comment (humphd) [inline src/utils/DragAndDrop.js]: I think that's a good idea, actually.  @MargarytaChepiga could you add an `isArchive()` method to this file https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/content.js.  It could do basically the same thing you're doing here.  Also, you could then change all of these long checks for zip/tar to just be `Content.isArchive()`.  Make sense?

**Comment by humphd** [comment]
@MargarytaChepiga how are the changes to this coming?  Do you have questions or any issues?  It would be nice to get this landed soon.

I left you some notes in other bugs about the text missing from your dialogs.  Check your notifications/github emails.

**Comment by MargarytaChepiga** [comment]
@humphd  Thank you so much! Still working on it, should be able to finish it soon though.

File: src/utils/DragAndDrop.js (hunk 3 above)
Comment (humphd) [inline src/utils/DragAndDrop.js]: Here you can do `var ext = FilerUtils.normalizeExtension(Path.extname(item.name), true);` instead of all this code (NOTE: the second arg is `true` so we will exclude the period in this case, since `LanguageManager.getLanguageForExtension()` assumes there isn't one).

**Comment by MargarytaChepiga** [comment]
@humphd Should we also use FilerUtils.normalizeExtension on line 411 in DragAndDrop.js?
screenshot from 2017-03-29 16-01-25
So instead of:
var ext = Path.extname(filename).toLowerCase();
We could have:
var ext = FilerUtils.normalizeExtension(Path.extname(filename));
or 
var ext = FilerUtils.normalizeExtension(Path.extname(filename), false);

**Comment by humphd** [comment]
It won't need it, and you can remove the `.toLowerCase()` on line 411 since your new code will get called indirectly in `isUTF8Encoded`.

**Comment by MargarytaChepiga** [comment]
@humphd  Got it, Thank you!

**Comment by MargarytaChepiga** [comment]
I also changed the string in thimble: https://github.com/mozilla/thimble.mozilla.org/pull/1933

File: locales/en-US/editor.properties [hunk 4]
```diff
@@ -126,6 +126,7 @@ CMD_TOGGLE_QUICK_DOCS=Quick Docs
 # Drag and Drop
 
 DND_MAX_FILE_SIZE_EXCEEDED=file exceeds maximum supported size: 3MB
+DND_MAX_ARCHIVE_SIZE_EXCEEDED=file exceeds maximum supported size: 5MB
```
Comment (humphd) [inline locales/en-US/editor.properties]: Because these 2 strings are basically identical, I think we should try to reuse them, and just pass the number that differs at runtime:

* Let's create a new string, `DND_MAX_SIZE_EXCEEDED=file exceeds maximum supported size: {0} MB.`
* Everywhere that `Strings.DND_MAX_FILE_SIZE_EXCEEDED` is used, change it to `StringUtils.format(Strings.DND_MAX_SIZE_EXCEEDED, "3")`
* Everywhere that `DND_MAX_ARCHIVE_SIZE_EXCEEDED` is used, change it to `StringUtils.format(Strings.DND_MAX_SIZE_EXCEEDED, "5")`

You can see an example of this in:

* https://github.com/MargarytaChepiga/brackets/blob/d2d3cce9f1617137b0c1b9d7e79889c99e51ef24/locales/en-US/editor.properties#L11
* https://github.com/adobe/brackets/blob/f028b3ecdf2017498196f39ddbbe4a2364366c92/src/file/FileUtils.js#L190

File: src/filesystem/impls/filer/lib/content.js [hunk 5]
```diff
@@ -37,8 +38,13 @@ define(function (require, exports, module) {
...
         },
 
+        isArchive: function(ext) {
+            ext = FilerUtils.normalizeExtension(ext);
+            return ext === '.zip' || ext === '.tar'; 
+        },
+
         mimeFromExt: function(ext) {
-            ext = ext.toLowerCase();
+            ext = FilerUtils.normalizeExtension(ext, false);
```
Comment (humphd) [inline src/filesystem/impls/filer/lib/content.js]: You can just do `ext = FilerUtils.normalizeExtension(ext)` and not bother passing `false`, since passing nothing will also produce a `false` type value in the function.  This is more common in JS.

File: src/utils/DragAndDrop.js [hunk 6]
```diff
@@ -47,12 +47,16 @@ define(function (require, exports, module) {
         Content         = require("filesystem/impls/filer/lib/content"),
         LanguageManager = require("language/LanguageManager"),
         StartupState    = require("bramble/StartupState"),
-        ArchiveUtils    = require("filesystem/impls/filer/ArchiveUtils");
+        ArchiveUtils    = require("filesystem/impls/filer/ArchiveUtils"),
+        FilerUtils = require('filesystem/impls/filer/FilerUtils');
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: Line up the `=` with the lines above please.

**Comment by humphd** [comment]
@gideonthomas I really want to get this landed ASAP, since my dnd stuff is blocked on it.  Can you review, since you had issues with the localization strategy I suggested?

@Pomax that's probably something for another bug, if you want to file.

File: src/utils/DragAndDrop.js [hunk 7]
```diff
@@ -332,13 +336,19 @@ define(function (require, exports, module) {
          * or not a mime type we care about, reject it.
          */
         function rejectImport(item) {
-            if (item.size > byteLimit) {
-                return new Error(Strings.DND_MAX_FILE_SIZE_EXCEEDED);
+            var ext = FilerUtils.normalizeExtension(Path.extname(item.name), true);
+
+            if (item.size > archiveByteLimit) {
```
Comment (gideonthomas) [inline src/utils/DragAndDrop.js]: I Think this is incorrect:
let's change this logic to:
```js
var sizeLimit = Content.isArchive(ext) ? archiveByteLimit : byteLimit;
var sizeLimitMb = (sizeLimit / (1024 * 1024)).toString();

if (item.size > sizeLimit) {
  return new Error(StringUtils.format(Strings.DND_MAX_SIZE_EXCEEDED, sizeLimitMb));
}
```

**Comment by humphd** [comment]
@MargarytaChepiga I need this change to get merged so I can merge my dependent PR.  Please make the change @gideonthomas requested above.  Thanks.

**Comment by humphd** [comment]
NOTE: waiting on string change in https://github.com/mozilla/thimble.mozilla.org/pull/1933

File: locales/en-US/editor.properties [hunk 8]
```diff
@@ -125,7 +125,7 @@ CMD_TOGGLE_QUICK_DOCS=Quick Docs
 
 # Drag and Drop
 
-DND_MAX_FILE_SIZE_EXCEEDED=file exceeds maximum supported size: 3MB
+DND_MAX_SIZE_EXCEEDED=file exceeds maximum supported size: {0} MB.
```
Comment (humphd) [inline locales/en-US/editor.properties]: @gideonthomas can you clarify what we need to do here, now that this string is in Thimble?  Do we do both?

Comment (gideonthomas) [inline locales/en-US/editor.properties]: technically no. You can just have it in Thimble and it will automatically come into this repo

**Comment by humphd** [comment]
@MargarytaChepiga OK, so let's have you remove that string, and deal with whatever merge conflicts are in this branch (e.g., rebase and fix them).  Then we can merge this!

**Comment by MargarytaChepiga** [comment]
@humphd Just to clarify, you want me to remove this string from brackets: DND_MAX_SIZE_EXCEEDED=file exceeds maximum supported size: {0} MB. ?

**Comment by humphd** [comment]
Correct.  Remove it here (in Brackets), because it will get automatically pulled in from Thimble.  I know, it's overly complicated :(

File: src/filesystem/impls/filer/FilerUtils.js [hunk 9]
```diff
@@ -22,4 +22,11 @@ define(function (require, exports, module) {
             return path;
         }
     };
+
+    // Normalize '.html', 'html', '.HTML', 'HTML' all to '.html', unless exludePeriod is 'true'
+    // then make it just 'html' without the period.
+    exports.normalizeExtension = function(ext, exludePeriod) {
```
Comment (humphd) [inline src/filesystem/impls/filer/FilerUtils.js]: `exludePeriod` should be `excludePeriod` (missing a `c`).  Can you please fix here and in the line below?

File: src/utils/DragAndDrop.js [hunk 10]
```diff
@@ -332,13 +336,16 @@ define(function (require, exports, module) {
          * or not a mime type we care about, reject it.
          */
         function rejectImport(item) {
-            if (item.size > byteLimit) {
-                return new Error(Strings.DND_MAX_FILE_SIZE_EXCEEDED);
-            }
+            var ext = FilerUtils.normalizeExtension(Path.extname(item.name), true);
```
Comment (humphd) [inline src/utils/DragAndDrop.js]: You don't need to call `FilerUtils.normalizeExtension()` here, since you already call it in `Content.isArchive()`.  Basically, we need it, but we only need it once.

**Comment by MargarytaChepiga** [comment]
@humphd Please review the changes. Thanks.

**Comment by gideonthomas** [comment]
@MargarytaChepiga can you rebase your branch onto mozilla/master please?

**Comment by humphd** [comment]
Rebased and fixed the `locale/` change myself, merged in https://github.com/mozilla/brackets/commit/a47ee0def94f76390aaaf87e5b6fbf4093807684.  Closing.
